File: day03/part2.ts
import { run } from "../common"
import { parseInput } from "./input";

const BATTERY_COUNT = 12;

function solveLine(digits: number[]): number {
    const chosenBatteries: number[] = [];
    for(let i = BATTERY_COUNT - 1; i >= 0; i--) {
        const next = digits.slice(0, digits.length - i)
            .sort()
            .reverse()
            .at(0)!;
        const nextIdx = digits.indexOf(next);
        digits = digits.slice(nextIdx + 1);
        chosenBatteries.push(next);
    }
    return parseInt(chosenBatteries.join(""), 10);
}
function solve(lines: number[][]): number {
    return lines
        .map(solveLine)
        .reduce((acc, val) => acc + val, 0);
}

await run(parseInput, solve);

File: day06/part2.rs
use common::run;

use crate::input::{Expression, Input, Operation};

mod input;

fn split_at_indices<'a>(line: &'a str, indices: &[usize]) -> impl Iterator<Item = &'a str> {
    indices
        .iter()
        .enumerate()
        .map(|(i, idx)| {
            let end = if i < indices.len() - 1 {
                indices[i + 1] - 1
            } else {
                line.len()
            };
            *idx..end
        })
        .map(|range| line.get(range).unwrap())
}

pub fn parse_input(input_str: &str) -> Input {
    let mut line_iter = input_str.lines();
    let op_line = line_iter.next_back().expect("Could not take operator line");
    let (op_indices, operations): (Vec<_>, Vec<_>) = op_line
        .chars()
        .enumerate()
        .filter(|(_, c)| !c.is_ascii_whitespace())
        .map(|(i, c)| (i, Operation::try_from(c).unwrap()))
        .unzip();

    let mut line_iters = line_iter
        .map(|line| split_at_indices(line, &op_indices))
        .collect::<Vec<_>>();

    operations
        .into_iter()
        .map(|operation| {
            let lines = line_iters
                .iter_mut()
                .map(Iterator::next)
                .map(Option::unwrap)
                .collect::<Vec<_>>();

            // All lines have the same length.
            let max_len = lines[0].len();

            let inputs = (0..max_len)
                .rev()
                .map(|idx| {
                    let digits = lines
                        .iter()
                        .map(|line| line.get(idx..(idx + 1)).unwrap_or(""))
                        .filter(|c| c != &" ")
                        .collect::<Vec<_>>()
                        .join("");
                    digits.parse().unwrap()
                })
                .collect();

            Expression { operation, inputs }
        })
        .collect()
}

fn solve_part2(input: Input) -> i64 {
    input
        .into_iter()
        .map(|expr| {
            expr.inputs
                .into_iter()
                .reduce(|acc, val| expr.operation.call(acc, val))
                .unwrap()
        })
        .sum::<usize>() as i64
}

fn main() {
    run(parse_input, solve_part2);
}
